components/brave_wallet/browser/simple_hash_client.cc

```diff
@@ -245,6 +252,259 @@ void SimpleHashClient::OnFetchAllNFTsFromSimpleHash(
+  }
+
+  // For each NFT identifier, create the NftIdentifier from the corresponding
+  // chain_id, contract_address, and token_id (if applicable), and look up the
+  // map of owners. Check if the wallet_address is in the owners map and add
+  // the balance to the balances vector (keeping the original order).
+  std::vector<uint64_t> balances;
+  for (const auto& nft_identifier : nft_identifiers) {
+    auto it = owners->find(nft_identifier);
+    if (it == owners->end()) {
+      balances.push_back(0);
+      continue;
+    }
+
+    auto owner_it = it->second.find(wallet_address);
+    if (owner_it == it->second.end()) {
+      balances.push_back(0);
+      continue;
+    }
+
+    balances.push_back(owner_it->second);
+  }
+
+  std::move(callback).Run(std::move(balances));
+}
+
+void SimpleHashClient::GetNftMetadatas(
+    mojom::CoinType coin,
+    std::vector<mojom::NftIdentifierPtr> nft_identifiers,
+    GetNftMetadatasCallback callback) {
+  if (nft_identifiers.size() > kSimpleHashMaxBatchSize) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  GURL url = SimpleHashClient::GetNftsUrl(coin, nft_identifiers);
+  if (!url.is_valid()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  auto internal_callback = base::BindOnce(
+      &SimpleHashClient::OnGetNftsForMetadatas, weak_ptr_factory_.GetWeakPtr(),
+      coin, std::move(nft_identifiers), std::move(callback));
+
+  api_request_helper_->Request("GET", url, "", "", std::move(internal_callback),
+                               MakeBraveServicesKeyHeaders(),
+                               {.auto_retry_on_network_change = true});
+}
+
+void SimpleHashClient::OnGetNftsForMetadatas(
+    mojom::CoinType coin,
+    std::vector<mojom::NftIdentifierPtr> nft_identifiers,
+    GetNftMetadatasCallback callback,
+    APIRequestResult api_request_result) {
+  if (!api_request_result.Is2XXResponseCode() ||
+      api_request_result.value_body().is_none()) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  // A map of composite token IDs to their metadata
+  std::optional<base::flat_map<std::string, mojom::NftMetadataPtr>> metadatas =
+      ParseMetadatas(api_request_result.TakeBody());
+
+  // For each contract address, create the composite key from the corresponding
+  // chain_id and contract_address at the same index, and look up the metadata
+  // in the map using the case insensitive composite key and add it to the token
+  // urls vector (keeping the original order).
+  std::vector<mojom::NftMetadataPtr> nft_metadatas;
+  for (size_t i = 0; i < nft_identifiers.size(); i++) {
+    std::string composite_key;
+    if (coin == mojom::CoinType::ETH) {
+      composite_key = nft_identifiers[i]->chain_id + "." +
+                      nft_identifiers[i]->contract_address + "." +
+                      nft_identifiers[i]->token_id;
+    } else {
+      composite_key = nft_identifiers[i]->chain_id + "." +
+                      nft_identifiers[i]->contract_address;
+    }
+
```

Comment:
We don't need conept of string composite_key when there is a nice struct mojom::NftIdentifierPtr. Parser above should return map<NftIdentifier, NftMetadata> which also looks natural.